core/grants.go

```diff
@@ -0,0 +1,194 @@
+package core
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+
+	chainjson "chain/encoding/json"
+	"chain/errors"
+	"chain/net/http/authz"
+	"chain/net/http/httpjson"
+)
+
+// an api-friendly representation of a grant
+type apiGrant struct {
+	GuardType string        `json:"guard_type"`
+	GuardData chainjson.Map `json:"guard_data"`
+	Policy    string        `json:"policy"`
+	CreatedAt string        `json:"created_at"`
+}
+
+// ErrMissingTokenID is returned when a token does not exist.
+var errMissingTokenID = errors.New("id does not exist")
+
+func (a *API) createGrant(ctx context.Context, x apiGrant) error {
+	guardData, err := x.GuardData.MarshalJSON()
+	if err != nil {
+		// chainjson.Map implementation means this should never happen ¯\_(ツ)_/¯
+		return errors.Wrap(err)
+	}
+
+	// before we go any further, make sure this token is real
+	var gd map[string]string
+	err = json.Unmarshal(guardData, &gd)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	id, ok := gd["id"]
+	if !ok || !a.accessTokens.CheckID(ctx, id) {
+		return errMissingTokenID
+	}
+
+	g := authz.Grant{
+		GuardType: x.GuardType,
+		GuardData: guardData,
+		Policy:    x.Policy,
+		CreatedAt: time.Now().UTC().Format(time.RFC3339),
+	}
+
+	data, err := a.raftDB.Get(ctx, grantPrefix+x.Policy)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	if data == nil {
+		// if there aren't any grants associated with this policy, go ahead
+		// and chuck this into raftdb
+		gList := &authz.GrantList{
+			Grants: []*authz.Grant{&g},
+		}
+		val, err := proto.Marshal(gList)
+		if err != nil {
+			return errors.Wrap(err)
+		}
+		err = a.raftDB.Set(ctx, grantPrefix+x.Policy, val)
```

Comment:
I suggest using Insert instead of Set here, so it will fail in case someone concurrently added another grant for this policy instead of clobbering the other value.